File: Equality.py
import IR
import Util
import TypeVariableInfo
import MemberInfo

class InferenceEngine(object):
    def __init__(self):
        self.fn = None
        self.next = 0
        self.tv_info_vars = {}
        self.substitution = TypeVariableInfo.Substitution()

    def inferFn(self, fn):
        self.fn = fn
        #print("Equality for %s" % fn.name)
        self.initialize()
        self.mergeInstructions()
        self.mergeMembers()
        self.finalize()
        #self.dump()

    def nextOwnershipVar(self):
        n = self.next
        self.next += 1
        v = TypeVariableInfo.OwnershipVar()
        v.value = n
        return v

    def nextGroupVar(self):
        n = self.next
        self.next += 1
        v = TypeVariableInfo.GroupVar()
        v.value = n
        return v

    def nextTypeVariableInfo(self):
        tv_info = TypeVariableInfo.TypeVariableInfo()
        tv_info.ownership_var = self.nextOwnershipVar()
        tv_info.group_var = self.nextGroupVar()
        return tv_info

    def initialize(self):
        for arg in self.fn.args:
            tv_info = self.nextTypeVariableInfo()
            self.tv_info_vars[arg.name] = tv_info
        for block in self.fn.body.blocks:
            for i in block.instructions:
                i.tv_info = self.nextTypeVariableInfo()
                if isinstance(i, IR.Bind):
                    tv_info = self.nextTypeVariableInfo()
                    self.tv_info_vars[i.name] = tv_info
                if isinstance(i, IR.ValueRef):
                    root = self.nextGroupVar()
                    for index in i.indices:
                        member_info = MemberInfo.MemberInfo()
                        member_info.root = root
                        member_info.kind = MemberInfo.MemberKind()
                        member_info.kind.type = "field"
                        member_info.kind.index = index
                        member_info.info = self.nextTypeVariableInfo()
                        root = member_info.info.group_var
                        i.members.append(member_info)
                    if len(i.members) != 0:
                        i.members[-1].info.group_var = i.tv_info.group_var
                if isinstance(i, IR.NamedFunctionCall):
                    if i.ctor:
                        for (index, arg) in enumerate(i.args):
                            member_info = MemberInfo.MemberInfo()
                            member_info.root = i.tv_info.group_var
                            member_info.kind = MemberInfo.MemberKind()
                            member_info.kind.type = "field"
                            member_info.kind.index = index
                            member_info.info = self.nextTypeVariableInfo()
                            i.members.append(member_info)

    def unifyOwnership(self, o1, o2):
        o1 = self.substitution.applyOwnershipVar(o1)
        o2 = self.substitution.applyOwnershipVar(o2)
        self.substitution.addOwnershipVar(o1, o2)

    def unifyGroup(self, g1, g2):
        g1 = self.substitution.applyGroupVar(g1)
        g2 = self.substitution.applyGroupVar(g2)
        self.substitution.addGroupVar(g1, g2)

    def unify(self, info1, info2):
        self.unifyOwnership(info1.ownership_var, info2.ownership_var)
        self.unifyGroup(info1.group_var, info2.group_var)

    def unifyInstrAndVar(self, id, name):
        info1 = self.getInstructionTypeVariableInfo(id)
        info2 = self.tv_info_vars[name]
        self.unify(info1, info2)

    def unifyInstrs(self, id1, id2):
        info1 = self.getInstructionTypeVariableInfo(id1)
        info2 = self.getInstructionTypeVariableInfo(id2)
        self.unify(info1, info2)

    def getInstructionTypeVariableInfo(self, id):
        return self.fn.body.getInstruction(id).tv_info

    def processBlock(self, block):
        for i in block.instructions:
            if isinstance(i, IR.Bind):
                self.unifyInstrAndVar(i.rhs, i.name)
            elif isinstance(i, IR.NamedFunctionCall):
                if i.ctor:
                    for (index, arg) in enumerate(i.args):
                        member_info = i.members[index]
                        arg_info = self.getInstructionTypeVariableInfo(arg)
                        self.unify(arg_info, member_info.info)
            elif isinstance(i, IR.ValueRef):
                if len(i.members) == 0:
                    self.unifyGroup(self.tv_info_vars[i.name].group_var, i.tv_info.group_var)    
                else:
                    self.unifyGroup(self.tv_info_vars[i.name].group_var, i.members[0].root)
            elif isinstance(i, IR.BoolLiteral):
                pass
            elif isinstance(i, IR.If):
                true_branch = self.fn.body.getBlock(i.true_branch)
                false_branch = self.fn.body.getBlock(i.false_branch)
                self.processBlock(true_branch)
                self.processBlock(false_branch)
                t_id = true_branch.getLastReal().id
                f_id = false_branch.getLastReal().id
                self.unifyInstrs(t_id, f_id)
            elif isinstance(i, IR.DropVar):
                pass
            elif isinstance(i, IR.BlockRef):
                b = self.fn.body.getBlock(i.value)
                self.processBlock(b)
                l_id = b.getLastReal().id
                self.unifyInstrs(i.id, l_id)
            elif isinstance(i, IR.Converter):
                info1 = self.getInstructionTypeVariableInfo(i.id)
                info2 = self.getInstructionTypeVariableInfo(i.arg)
                self.unifyGroup(info1.group_var, info2.group_var)
            elif isinstance(i, IR.Nop):
                pass
            else:
                Util.error("OI: grouping not handling %s %s" % (type(i), i))

    def mergeInstructions(self):
        block = self.fn.body.getFirst()
        self.processBlock(block)
    
    def mergeMembers(self):
        while True:
            member_map = {}
            for block in self.fn.body.blocks:
                for i in block.instructions:
                    for member in i.members:
                        member.root = self.substitution.applyGroupVar(member.root)
                        member_map[(member.root, member.kind.index)] = []
            
            for block in self.fn.body.blocks:
                for i in block.instructions:
                    for member in i.members:
                        member.info = self.substitution.applyTypeVariableInfo(member.info)
                        member_map[(member.root, member.kind.index)].append(member.info)
        
            unified = False
            for entries in member_map.values():
                entries = list(set(entries))
                if len(entries) > 1:
                    first = entries[0]
                    first = self.substitution.applyTypeVariableInfo(first)
                    for entry in entries:
                        entry = self.substitution.applyTypeVariableInfo(entry)
                        self.unify(first, entry)
                        unified = True
            if not unified:
                break

    def finalize(self):
        for block in self.fn.body.blocks:
            for i in block.instructions:
                for member in i.members:
                    member.root = self.substitution.applyGroupVar(member.root)
                    member.info = self.substitution.applyTypeVariableInfo(member.info)
                i.tv_info = self.substitution.applyTypeVariableInfo(i.tv_info)
    
    def dump(self):
        for block in self.fn.body.blocks:
            print("#%s block" % block.id)
            for i in block.instructions:
                print("%5s %35s - %4s %s" % (i.id, i, i.tv_info, i.members))

def infer(program):
    for f in program.functions.values():
        engine = InferenceEngine()
        engine.inferFn(f)

File: Siko/src/siko/ownershipold/Equality.rs
use std::collections::BTreeMap;

use crate::siko::{
    ir::Function::{Block, Function, InstructionId, InstructionKind, ValueKind},
    ownership::MemberInfo::{MemberInfo, MemberKind},
    qualifiedname::QualifiedName,
};

use super::{
    dataflowprofile::{
        DataFlowProfile::DataFlowProfile, DataFlowProfileStore::DataFlowProfileStore,
    },
    Instantiator::Instantiator,
    OwnershipInferenceInfo::OwnershipInferenceInfo,
    OwnershipInferenceInfo::TypedId,
    TypeVariableInfo::{GroupTypeVariable, OwnershipTypeVariable, Substitution, TypeVariableInfo},
};

pub struct EqualityEngine<'a> {
    function: &'a Function,
    profileStore: &'a DataFlowProfileStore,
    substitution: Substitution,
    groupProfiles: &'a BTreeMap<QualifiedName, DataFlowProfile>,
    profiles: BTreeMap<InstructionId, DataFlowProfile>,
    ownershipInferenceInfo: OwnershipInferenceInfo,
}

impl<'a> EqualityEngine<'a> {
    pub fn new(
        function: &'a Function,
        profileStore: &'a DataFlowProfileStore,
        groupProfiles: &'a BTreeMap<QualifiedName, DataFlowProfile>,
    ) -> EqualityEngine<'a> {
        EqualityEngine {
            function: function,
            profileStore: profileStore,
            substitution: Substitution::new(),
            groupProfiles: groupProfiles,
            profiles: BTreeMap::new(),
            ownershipInferenceInfo: OwnershipInferenceInfo::new(),
        }
    }

    pub fn run(&mut self, initializeSignature: bool) {
        self.initialize(initializeSignature);
        let block = self.function.getFirstBlock();
        self.processBlock(block); // TODO: fix this, no clone!
    }

    fn initialize(&mut self, initializeSignature: bool) {
        if initializeSignature {
            for param in &self.function.params {
                let tvInfo = self.ownershipInferenceInfo.nextTypeVariableInfo();
                self.ownershipInferenceInfo
                    .tvInfos
                    .insert(TypedId::Value(param.getName()), tvInfo.clone());
                self.ownershipInferenceInfo.signature.args.push(tvInfo);
            }
        }
        for block in &self.function.body.as_ref().unwrap().blocks {
            for instruction in &block.instructions {
                let tvInfo = self.ownershipInferenceInfo.nextTypeVariableInfo();
                self.ownershipInferenceInfo
                    .tvInfos
                    .insert(TypedId::Instruction(instruction.id), tvInfo);
                match &instruction.kind {
                    InstructionKind::ValueRef(_, _, indices) => {
                        let mut members = Vec::new();
                        let mut root = self.ownershipInferenceInfo.nextGroupVar();
                        for index in indices {
                            let member = MemberInfo::new(
                                root,
                                MemberKind::Field,
                                *index,
                                self.ownershipInferenceInfo.nextTypeVariableInfo(),
                            );
                            root = member.info.group;
                            members.push(member);
                        }
                        if !members.is_empty() {
                            members.last_mut().unwrap().info.group = tvInfo.group;
                            self.ownershipInferenceInfo
                                .members
                                .insert(instruction.id, members);
                        }
                    }
                    InstructionKind::Bind(name, _) => {
                        let tvInfo = self.ownershipInferenceInfo.nextTypeVariableInfo();
                        self.ownershipInferenceInfo
                            .tvInfos
                            .insert(TypedId::Value(name.clone()), tvInfo);
                    }
                    _ => {}
                }
            }
        }
    }

    fn getInstructionTypeVariableInfo(&self, id: InstructionId) -> TypeVariableInfo {
        self.ownershipInferenceInfo
            .tvInfos
            .get(&TypedId::Instruction(id))
            .unwrap()
            .clone()
    }

    fn getValueTypeVariableInfo(&self, v: String) -> TypeVariableInfo {
        self.ownershipInferenceInfo
            .tvInfos
            .get(&&TypedId::Value(v))
            .unwrap()
            .clone()
    }

    fn unifyOwnership(&mut self, o1: OwnershipTypeVariable, o2: OwnershipTypeVariable) {
        let o1 = self.substitution.applyOwnershipVar(o1);
        let o2 = self.substitution.applyOwnershipVar(o2);
        self.substitution.addOwnershipVar(o1, o2);
    }

    fn unifyGroup(&mut self, g1: GroupTypeVariable, g2: GroupTypeVariable) {
        let g1 = self.substitution.applyGroupVar(g1);
        let g2 = self.substitution.applyGroupVar(g2);
        self.substitution.addGroupVar(g1, g2);
    }

    fn unify(&mut self, info1: TypeVariableInfo, info2: TypeVariableInfo) {
        self.unifyOwnership(info1.owner, info2.owner);
        self.unifyGroup(info1.group, info2.group);
    }

    fn processBlock(&mut self, block: &Block) {
        for instruction in &block.instructions {
            match &instruction.kind {
                InstructionKind::FunctionCall(name, args) => {
                    let profile;
                    if let Some(p) = self.groupProfiles.get(name) {
                        profile = Some(p.clone());
                    } else {
                        profile = self.profileStore.getProfile(name);
                    }
                    if let Some(profile) = profile {
                        let mut instantiator = Instantiator::new(
                            self.ownershipInferenceInfo.signature.allocator.clone(),
                        );
                        let profile = instantiator.instantiateProfile(profile);
                        self.ownershipInferenceInfo.signature.allocator = instantiator.allocator;
                        let resInfo = self.getInstructionTypeVariableInfo(instruction.id);
                        for path in &profile.paths {
                            let argId = args[path.index as usize];
                            let argInfo = self.getInstructionTypeVariableInfo(argId);
                            self.unify(argInfo, path.arg);
                        }
                        for (index, argId) in args.iter().enumerate() {
                            let sigArgInfo = profile.signature.args[index];
                            let argInfo = self.getInstructionTypeVariableInfo(*argId);
                            self.unify(argInfo, sigArgInfo);
                        }
                        self.unify(resInfo, profile.signature.result);
                        self.profiles.insert(instruction.id, profile);
                    }
                }
                InstructionKind::DynamicFunctionCall(_, _) => {
                    panic!("dynamic function call in equality")
                }
                InstructionKind::If(_, trueBranch, falseBranch) => {
                    let trueBlock = self.function.getBlockById(*trueBranch);
                    self.processBlock(trueBlock);
                    if let Some(falseBranch) = falseBranch {
                        let falseBlock = self.function.getBlockById(*falseBranch);
                        self.processBlock(falseBlock);
                        self.unify(
                            self.getInstructionTypeVariableInfo(trueBlock.getLastId()),
                            self.getInstructionTypeVariableInfo(falseBlock.getLastId()),
                        );
                    }
                }
                InstructionKind::BlockRef(id) => {
                    let block = self.function.getBlockById(*id);
                    self.processBlock(block);
                    self.unify(
                        self.getInstructionTypeVariableInfo(block.getLastId()),
                        self.getInstructionTypeVariableInfo(instruction.id),
                    );
                }
                InstructionKind::Loop(_, _, _) => todo!(),
                InstructionKind::ValueRef(v, _, _) => {
                    if let Some(members) = self.ownershipInferenceInfo.members.get(&instruction.id)
                    {
                        if let Some(name) = v.getValue() {
                            self.unifyGroup(
                                self.getValueTypeVariableInfo(name).group,
                                members[0].root,
                            );
                        } else {
                            match v {
                                ValueKind::Implicit(id) => {
                                    self.unifyGroup(
                                        self.getInstructionTypeVariableInfo(*id).group,
                                        members[0].root,
                                    );
                                }
                                _ => unreachable!(),
                            }
                        }
                    } else {
                        if let Some(name) = v.getValue() {
                            self.unifyGroup(
                                self.getValueTypeVariableInfo(name).group,
                                self.getInstructionTypeVariableInfo(instruction.id).group,
                            );
                        } else {
                            match v {
                                ValueKind::Implicit(id) => {
                                    self.unifyGroup(
                                        self.getInstructionTypeVariableInfo(*id).group,
                                        self.getInstructionTypeVariableInfo(instruction.id).group,
                                    );
                                }
                                _ => unreachable!(),
                            }
                        }
                    }
                }
                InstructionKind::Bind(name, rhs) => self.unify(
                    self.getValueTypeVariableInfo(name.clone()),
                    self.getInstructionTypeVariableInfo(*rhs),
                ),
                InstructionKind::Tuple(_) => todo!(),
                InstructionKind::TupleIndex(_, _) => todo!(),
                InstructionKind::StringLiteral(_) => todo!(),
                InstructionKind::IntegerLiteral(_) => todo!(),
                InstructionKind::CharLiteral(_) => todo!(),
                InstructionKind::Continue(_, _) => todo!(),
                InstructionKind::Break(_, _) => todo!(),
                InstructionKind::Return(_) => todo!(),
                InstructionKind::Ref(_) => todo!(),
            }
        }
        // for i in block.instructions:
        //     if isinstance(i, Instruction.Bind):
        //         self.unifyInstrAndVar(i.rhs, i.name)
    }

    pub fn dump(&self) {
        println!("Sig: {:?}", self.ownershipInferenceInfo.signature);
        for profile in self.profiles.values() {
            println!("Profile sig: {:?}", profile.signature);
            println!("Paths {:?}", profile.paths);
        }
        for block in &self.function.body.as_ref().unwrap().blocks {
            println!("#%s block {}", block.id);
            for i in &block.instructions {
                let tvInfo = self
                    .ownershipInferenceInfo
                    .tvInfos
                    .get(&TypedId::Instruction(i.id))
                    .unwrap();
                if let Some(members) = self.ownershipInferenceInfo.members.get(&i.id) {
                    println!("{} - {} {:?}", i, tvInfo, members);
                } else {
                    println!("{} - {}", i, tvInfo);
                }
            }
        }
    }
}
